### ISSUE TITLE
IBM - Undefined values in ypraf in case of periodicity

### ISSUE THREAD

**Issue by mathrack** [xcompact3d/Incompact3d#293]
The size of the array `ypraf` is `nyraf = nym nraf`. In case of periodicity, `nym=ny`. Otherwise `nym=ny-1`.

The loop used to define `ypraf` will populate the array from `1` to `(ny-1) nraf`.

https://github.com/xcompact3d/Incompact3d/blob/46e874e236265c2f2671f254a9e3f4415d1e40d0/src/genepsi3d.f90#L228

In case of periodicity, the array will contain `nraf` undefined values at the end. This is an issue because it can lead to floating point exceptions when computing `yepsi`

https://github.com/xcompact3d/Incompact3d/blob/46e874e236265c2f2671f254a9e3f4415d1e40d0/src/genepsi3d.f90#L233

When the mesh is regular, one can simply defined `ypraf` using
```
    do j = 1, nyraf
          ypraf(j) = (j-1) * dyraf
    enddo
```

In case of stretching, I think one would have to use a subroutine similar to stretching

https://github.com/xcompact3d/Incompact3d/blob/46e874e236265c2f2671f254a9e3f4415d1e40d0/src/tools.f90#L1215

As a temporary workaround, I suggest to implement the simple loop to fix the issue in case of periodicity without stretching. In case of stretching combined with periodicity, we could keep the current code, print a warning and populate the end of the array `ypraf` with `yly + (j-(ny-1)*nraf)*dyraf` or something similar.

**Comment by slaizet** [comment]
okay, I find some time to look into this one. What about changing slightly the stretching subroutine so that it has 3 arguments: N (domain size), YP (stretching on the velocity mesh), YPI (stretching on the pressure mesh). This way, we could call the subroutine to define the stretching on the refined mesh. We would then just need to have
```
if (ncly) then
       call stretching(nyraf, ypraf,ypiraf)
    else
       do j = 1, nyraf
          ypraf(j) = (j-1) * dyraf
        enddo
    endif
```

**Comment by mathrack** [comment]
If the code supports periodicity + stretching, we could fix the IBM issue with a modification of the stretching subroutine. However, this looks a bit complex and the `poisson_000` subroutine seems unable to use a stretched mesh. A simple workaround could be 
```
if (ncly) then
   do j = 1, nyraf
      ypraf(j) = (j-1) * dyraf
   enddo
else
   ! Use the current code
endif
```
Somewhere after reading the input file, we could add a check to abort the simulation in case of periodicity + stretching

**Comment by slaizet** [comment]
The code does not support yet periodicity+stretching, but it will eventually :-) Anyway, even with non-periodic condition, the refinement procedure for the refined mesh is not great, hence why I think we should change the stretching subroutine.

**Comment by mathrack** [comment]
Ok, I will start a draft for a fix shortly !

**Comment by mathrack** [comment]
Fixed in  #314
